### PR TITLE
Add reporting related services to create and download them

### PR DESCRIPTION
**Pull request by mevinbabuc**
- Have added two more services to help with retrieving reports and downloading the report (in default CSV format)
- Have added a Report model as per comments, which has a download method to download reports.

```Python
from appnexus import connect, Report

connect("username", "password", test=True)

json = {
    "report_type": "insertion_order_frequency",
    "columns": [
        "hour",
        "imps",
        "clicks",
        "total_convs",
        "ctr",
        "convs_rate"
    ],
    "report_interval": "lifetime",
    "format": "csv"
}
create_report_response = Report(json).save()
Report.download(id=create_report_response['report_id'])
```

### CODE REVIEW

File: appnexus/client.py [hunk 1]
```diff
@@ -84,15 +84,22 @@ def _send(self, send_method, service, data=None, **kwargs):
...
-            response_data = response.json()
             try:
-                self.check_errors(response, response_data["response"])
-            except RateExceeded:
-                self._handle_rate_exceeded(response)
-            except NoAuth:
-                self.update_token()
-            else:
+                response_data = response.json()
+            except ValueError:
```
Comment (ramnes) [inline appnexus/client.py]: Can't we rather check the `Content-Type` or any other header information?

File: appnexus/client.py [hunk 2]
```diff
@@ -84,15 +84,22 @@ def _send(self, send_method, service, data=None, **kwargs):
...
+                # Must be a CSV or some other form of data, which is not JSON
                 valid_response = True
+                if response.status_code >= 200:
+                    return response.content
+            else:
+                try:
```
Comment (ramnes) [inline appnexus/client.py]: Please avoid `try` inside an `else`. I'm pretty sure a simpler structure could be used there, for example by using a bare `raise` if we're not entering into the previous `if`. That would also simplify the diff of that commit.

**Comment by ramnes** [comment]
Could you please update your example so that it actually works? What's `client.report_download`? Also, I'd like to see how you get the id rather than entering it manually.

**Comment by ramnes** [comment]
Okay so, following the discussion on #21, rather than implementing a `get` method inside `Model`, which as explained is not something we want to do, I'd like us to have something in the lines of:

```python
class Report(Model):
    def download(self):
        self.client.get("ReportDownload", self.id)
```
(obviously I didn't test this and it probably won't work out of the box)
 
so that we could simply do:
```python
# Create a report
report = Report(json).save()

# Download the report
report.download()
```

I feel that it's a much better interface. How do that sound to you? Can you make it that way?

**Comment by ramnes** [comment]
Then we would also remove `ReportDownload` from the `services_list`.

**Comment by mevinbabuc** [comment]
If you want to achieve a syntax close to what you showed above; we'll have to change the save method on the model to return a model object rather than result.

```python
    def save(self, **kwargs):
        payload = self.__dict__
        if "id" not in self.__dict__:
            logger.info("creating a {}".format(self.service))
            result = self.create(payload, **kwargs)
        else:
            result = self.modify(payload, id=self.id, **kwargs)
        return type(self)(result)
```

returning a `type(self)(result)` can achieve that. Then we can modify the Report Model to something like this

```python
class Report(Model):

    def download(self, **kwargs):
        return self.client.get("report-download", id=self.report_id)
```

with which you'll be able to do 

```
# Create a report
report = Report(json).save()

# Download the report
report.download()
```

Is the change to the save method to return a model object welcomed ?

**Comment by mevinbabuc** [comment]
I have the PR for that change ready; Let me know and I can raise one with that change

**Comment by ramnes** [comment]
Rather than instantiating a new object, you can simply use `Thingy.update` and do `self.update(result)`.

But yes, definitely, that would be awesome, please open that other PR! :+1: 

Looks like this is something we forgot when migrating to Thingy.

**Pull request closed without merging**